tests/test_visitor.py

```diff
@@ -845,7 +840,7 @@ def test_transformer_multinode_keys(frontend):
                 body=tuple(a.clone() for a in assigns))
     # Need to use NestedTransformer here, since replacement contains
     # the original nodes.
-    transformed = NestedTransformer({assigns: loop}).visit(routine.body)
+    transformed = NestedTransformer({assigns[0]: loop, assigns[1]: None}).visit(routine.body)
```

Comment:
This really does change the nature of the test (sub-tuple detection). Is this `Transformer` feature truly dependent on the source property of sub-expressions? The behaviour of transformers should not really be influenced by `expr.source` property, so I'm not sure why this is necessary?

Reply:
In the `Transformer._rebuild` method, before invalidating a node's `source` we check if its children have a `source` property. After removing `Expression.source` the children of an `Assignment` no longer have any source. So that test now passes and we invalidate `Assignment.source` in a transformer pass whereas previously we did not. The reason we don't hit this problem when the mapper key is a node rather than a tuple is that in the former case, we do the `is o in self.mapper` check before rebuilding `o`.
The cleanest way I can think to resolve this: in a `Transformer` pass, if `self.invalidate_source` is true we invalidate the source of mapper keys.